File: usvm-python/usvm-python-main/src/main/java/org/usvm/interpreter/ConcolicRunContext.java
package org.usvm.interpreter;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.usvm.machine.MockHeader;
import org.usvm.machine.PyContext;
import org.usvm.machine.PyState;
import org.usvm.machine.interpreters.symbolic.operations.tracing.PathDiversionException;
import org.usvm.machine.interpreters.symbolic.operations.tracing.SymbolicHandlerEvent;
import org.usvm.machine.model.PyModelHolder;
import org.usvm.machine.symbolicobjects.rendering.PyValueBuilder;
import org.usvm.machine.symbolicobjects.rendering.PyValueRenderer;
import org.usvm.machine.types.PythonTypeSystem;
import org.usvm.machine.utils.PythonMachineStatisticsOnFunction;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;


public class ConcolicRunContext {
    @Nullable
    public PyState curState;
    public PyContext ctx;
    public ArrayList<PyState> forkedStates = new ArrayList<>();
    public List<SymbolicHandlerEvent<Object>> pathPrefix;
    public MockHeader curOperation = null;
    public PyModelHolder modelHolder;
    public boolean allowPathDiversion;
    public PythonTypeSystem typeSystem;
    public PythonMachineStatisticsOnFunction statistics;
    public int maxInstructions;
    public int instructionCounter = 0;
    public boolean usesVirtualInputs = false;
    public Callable<Boolean> isCancelled;
    public PyValueBuilder builder = null;
    public PyValueRenderer renderer = null;

    public ConcolicRunContext(
        @NotNull PyState curState,
        PyContext ctx,
        PyModelHolder modelHolder,
        PythonTypeSystem typeSystem,
        boolean allowPathDiversion,
        PythonMachineStatisticsOnFunction statistics,
        int maxInstructions,
        Callable<Boolean> isCancelled
    ) {
        this.curState = curState;
        this.ctx = ctx;
        this.modelHolder = modelHolder;
        this.allowPathDiversion = allowPathDiversion;
        this.typeSystem = typeSystem;
        this.pathPrefix = curState.buildPathAsList();
        this.statistics = statistics;
        this.maxInstructions = maxInstructions;
        this.isCancelled = isCancelled;
    }

    public void pathDiversion() throws PathDiversionException {
        if (curState != null)
            curState.getMeta().setModelDied(true);
        if (allowPathDiversion) {
            curState = null;
        } else {
            throw new PathDiversionException();
        }
    }
}

File: usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/ConcolicRunContext.kt
package org.usvm.machine

import org.usvm.machine.interpreters.symbolic.operations.tracing.PathDiversionException
import org.usvm.machine.interpreters.symbolic.operations.tracing.SymbolicHandlerEvent
import org.usvm.machine.model.PyModelHolder
import org.usvm.machine.symbolicobjects.rendering.PyValueBuilder
import org.usvm.machine.symbolicobjects.rendering.PyValueRenderer
import org.usvm.machine.types.PythonType
import org.usvm.machine.types.PythonTypeSystem
import org.usvm.machine.utils.PythonMachineStatisticsOnFunction
import java.util.concurrent.Callable


class ConcolicRunContext(
    curState: PyState,
    val ctx: PyContext,
    val modelHolder: PyModelHolder,
    private val allowPathDiversion: Boolean,
    val statistics: PythonMachineStatisticsOnFunction,
    val maxInstructions: Int,
    val builder: PyValueBuilder,
    val renderer: PyValueRenderer,
    val isCancelled: Callable<Boolean>,
) {
    var curState: PyState?
    val forkedStates = mutableListOf<PyState>()
    var pathPrefix: List<SymbolicHandlerEvent<Any>>
    var instructionCounter = 0
    var usesVirtualInputs: Boolean = false

    val typeSystem: PythonTypeSystem
        get() = ctx.typeSystem<PythonType>() as PythonTypeSystem

    @JvmField
    var curOperation: MockHeader? = null

    init {
        this.curState = curState
        pathPrefix = curState.buildPathAsList()
    }

    @Throws(PathDiversionException::class)
    fun pathDiversion() {
        val state = curState
        if (state != null) {
            state.modelDied = true
        }
        curState = if (allowPathDiversion) {
            null
        } else {
            throw PathDiversionException()
        }
    }
}
